[J-Wang-CSMA] iP

J-wang-CSMA opened the pull request:
# Aurora

> "The happiness of your life depends upon the quality of your thoughts." – Marcus Aurelius [(source)](https://www.goodreads.com/quotes/139512-the-happiness-of-your-life-depends-upon-the-quality-of)

Aurora improves the quality of your thoughts by freeing your mind mundane tasks such as remembering your tasks while being an interesting companion. It is:
- user friendly
- easy to learn

All you need to do is,
1. download it from *[here](https://github.com/J-wang-CSMA/ip/releases)*.
2. add your tasks.
3. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Graphical interface
- [ ] Reminders (coming soon)
- [ ] Database integration (coming soon)

If you are a Java programmer, you can use it to practice Java too by trying to improve it. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Duke aurora1 = new Duke();
        aurora1.exeAurora();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

J-wang-CSMA committed:
Level 0. Rename, Greet, Exit

J-wang-CSMA committed:
Level 1. Echo

J-wang-CSMA committed:
Level 2. Add, List

J-wang-CSMA committed:
Level 3. Mark as Done

J-wang-CSMA committed:
Level 4.ToDos, Events, Deadlines

J-wang-CSMA committed:
Small bugfix with numbering

J-wang-CSMA committed:
A-TextUiTesting.

J-wang-CSMA committed:
A-TextUiTesting.

J-wang-CSMA committed:
Level 5. Handle Errors

J-wang-CSMA committed:
Level 6. Delete

J-wang-CSMA committed:
Additional testing after Increment 6

J-wang-CSMA committed:
Integrate Storage class with Duke for task persistence

J-wang-CSMA committed:
Merge branch 'branch-Level-7'

J-wang-CSMA committed:
Fix bug in Storage

J-wang-CSMA committed:
Fix bug in Storage class

J-wang-CSMA committed:
Merge branch 'branch-Level-7'

J-wang-CSMA committed:
Add DateTime integration

J-wang-CSMA committed:
Merge branch-Level-8 with master

J-wang-CSMA committed:
Add OOP features. Split Duke into Parser, Storage, TaskList, Commands.

J-wang-CSMA committed:
Add Package integration. Split program into different packages.

J-wang-CSMA committed:
Merge branch 'add-gradle-support'

J-wang-CSMA committed:
Add Gradle Support to main branch. Verify application runs with Gradle

J-wang-CSMA committed:
Add JUnit tests for objects. Pass all testcases for aurora.objects package

J-wang-CSMA committed:
Create jar executable

J-wang-CSMA committed:
Add Javadocs comments such that all public methods and classes have headers

J-wang-CSMA committed:
Change codestyle to fit coding standards

J-wang-CSMA committed:
Change codestyle to fit coding standards again

J-wang-CSMA committed:
Add Find feature to look for tasks using keywords

J-wang-CSMA committed:
Merge branch 'branch-Level-9'

J-wang-CSMA committed:
Merge branch 'branch-A-JavaDoc'

J-wang-CSMA committed:
Merge branch 'branch-A-CodingStandard'

J-wang-CSMA committed:
Move parsing logic to Parser class for better OOP

marcus-ny commented on src/main/java/aurora/parser/Parser.java:
```diff
@@ -0,0 +1,169 @@
+    public Command parseCommand(String command) {
+        String[] splitCommands = Parser.splitAtAllBlanks(command);
+        String mainC = splitCommands[0];
+        if (mainC.equalsIgnoreCase("bye")) {
```
Just a nitpick, but perhaps using switch cases here might be slightly neater.

marcus-ny commented on src/main/java/aurora/Duke.java:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Duke is the main class for the application that runs according to the commands given to it by the user.
+ */
+public class Duke {
```
Just a naming detail. I think that the class name should fit your chatbot's name Aurora, instead of the default "Duke". Similar case for your chatbot's exception class.

marcus-ny commented on src/main/java/aurora/Duke.java:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * Execution loop for commands.
+     */
+    public void executionLoop() {
```
Since this is a method, I think it would be better to use a verb (like executeLoop).

marcus-ny commented on src/main/java/aurora/ui/Ui.java:
```diff
@@ -0,0 +1,112 @@
+    /**
+     * Method to echo an add command.
+     */
+    public void echoAddTask(TaskList taskList) {
```
I like that you extracted this out rather than printing this along with the execution of adding a task! 💯

marcus-ny commented on src/main/java/aurora/objects/DukeException.java:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @return Message that describes the exception.
+     */
+    public String getExceptionMessage() {
```
I am not sure if you need an exception message for this class specifically if you already call the parent constructor with the same message, since you can get it with `Exception::getMessage`.

marcus-ny reviewed (commented):
Overall, naming conventions and code quality is very good with no blatant violations! 👍 Just a few nits to fix

J-wang-CSMA committed:
Implement JavaFX GUI for Aurora application.

J-wang-CSMA committed:
Add exit functionality to GUI

J-wang-CSMA committed:
Merge branch 'branch-Level-10'

LimZiJia commented on src/main/java/aurora/Duke.java:
```diff
@@ -0,0 +1,110 @@
+package aurora;
+
+import aurora.command.Command;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,116 @@
+package aurora.command;
+
+import aurora.objects.DukeException;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,116 @@
+        String[] descriptionAndDateSplit = Parser.splitAtFirstBlank(this.command);
+        if (descriptionAndDateSplit.length < 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter deadline, then specify the description of the task followed by the deadline.\n" +
```
This line is a little too long at 121 characters. The hard wrap limit is 120 according to the coding standard. Several of the same lines below have the same problem.

LimZiJia commented on src/main/java/aurora/command/DeleteCommand.java:
```diff
@@ -0,0 +1,96 @@
+package aurora.command;
+
+import aurora.objects.DukeException;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/command/DeleteCommand.java:
```diff
@@ -0,0 +1,96 @@
+        if (this.splitCommands.length != 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter unmark, then the number of the task you want to delete.");
+            // Solution adapted from https://www.baeldung.com/java-check-string-number
```
Good crediting, but  inline comments should be indented with the relative position in code. i.e. -4 spaces in this case. It just looks awkward otherwise.

LimZiJia commented on src/main/java/aurora/command/DeleteCommand.java:
```diff
@@ -0,0 +1,96 @@
+        if (this.splitCommands.length != 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter unmark, then the number of the task you want to delete.");
+            // Solution adapted from https://www.baeldung.com/java-check-string-number
```
Same issue. Also, instead of just referencing solution, can try to explain what you are trying to do with the inline comments also.

LimZiJia commented on src/main/java/aurora/command/EventCommand.java:
```diff
@@ -0,0 +1,136 @@
+        String[] descriptionAndDateSplit = Parser.splitAtFirstBlank(this.command);
+        if (descriptionAndDateSplit.length < 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter event, then specify the description of the task followed by the start and end " +
```
Line is too long, with many other copy pasted lines below also.

LimZiJia commented on src/main/java/aurora/command/FindCommand.java:
```diff
@@ -0,0 +1,88 @@
+            this.ui.printFoundList(this.foundList);
+        }
+    }
+    @Override
```
Nit picking but inconsistent line break here (no line break between methods)

LimZiJia commented on src/main/java/aurora/command/ListCommand.java:
```diff
@@ -0,0 +1,56 @@
+package aurora.command;
+
+import aurora.objects.Task;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/command/MarkCommand.java:
```diff
@@ -0,0 +1,101 @@
+package aurora.command;
+
+import aurora.objects.DukeException;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/command/MarkCommand.java:
```diff
@@ -0,0 +1,101 @@
+        if (this.splitCommands.length != 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter mark, then the number of the task you want to mark as done.");
+            // Solution adapted from https://www.baeldung.com/java-check-string-number
```
Same problem with other inline comments.

LimZiJia commented on src/main/java/aurora/command/MarkCommand.java:
```diff
@@ -0,0 +1,101 @@
+        if (this.splitCommands.length != 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter mark, then the number of the task you want to mark as done.");
+            // Solution adapted from https://www.baeldung.com/java-check-string-number
```
Same problem as other inline comments.

LimZiJia commented on src/main/java/aurora/command/TodoCommand.java:
```diff
@@ -0,0 +1,84 @@
+package aurora.command;
+
+import aurora.objects.DukeException;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,101 @@
+        if (this.splitCommands.length != 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter unmark, then the number of the task you want to unmark.");
+            // Solution adapted from https://www.baeldung.com/java-check-string-number
```
Same problem as other inline comments.

LimZiJia commented on src/main/java/aurora/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,101 @@
+        if (this.splitCommands.length != 2) {
+            throw new DukeException("Invalid number of arguments!\n" +
+                    "Make sure to enter unmark, then the number of the task you want to unmark.");
+            // Solution adapted from https://www.baeldung.com/java-check-string-number
```
Same problem as other inline comments.

LimZiJia commented on src/main/java/aurora/gui/DialogBox.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    /**
+     *
```
Missing JavaDoc

LimZiJia commented on src/main/java/aurora/gui/Main.java:
```diff
@@ -0,0 +1,50 @@
+package aurora.gui;
+
+import aurora.Duke;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/gui/Window.java:
```diff
@@ -0,0 +1,98 @@
+package aurora.gui;
+
+import aurora.Duke;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/parser/Parser.java:
```diff
@@ -0,0 +1,171 @@
+package aurora.parser;
+
+import aurora.command.ByeCommand;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/storage/Storage.java:
```diff
@@ -0,0 +1,154 @@
+package aurora.storage;
+
+import aurora.objects.Deadline;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/tasklist/TaskList.java:
```diff
@@ -0,0 +1,139 @@
+package aurora.tasklist;
+
+import aurora.objects.Deadline;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/ui/Ui.java:
```diff
@@ -0,0 +1,178 @@
+package aurora.ui;
+
+import aurora.objects.Task;
```
I think the checkstyle document they use have a different order of imports.
>The order rule "STATIC###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE" consists of:
      1. STATIC: static imports
      2. STANDARD_JAVA_PACKAGE: standard java/javax imports
      3. SPECIAL_IMPORTS: defined as org imports
      4. THIRD_PARTY_PACKAGE: defined as com imports

LimZiJia commented on src/main/java/aurora/ui/Ui.java:
```diff
@@ -0,0 +1,178 @@
+        int taskNumber = taskListArray.size();
+        printALine();
+        String echo =
+                "Added this task: \n" + taskListArray.get(taskNumber - 1) + "\nNumber of tasks in list: " + taskNumber;
```
Borderline acceptable length, but easily can just break line for better readability.

LimZiJia reviewed (commented):
Overall stuck to the coding standard well. They said to be consistent with the order of imports and to be fair it is consistent, but they also gave some order so better to follow I think.

You can also try to have more inline comments to explain some parts of your code.

J-wang-CSMA committed:
AuroraException and Aurora classes: Refactor DukeException
to AuroraException and Duke to Aurora for better readability.

Previously, AuroraException and Aurora were named DukeException
and Duke despite the application bearing the name Aurora. This
led to poor code readability.

Rename these classes to improve code quality.

J-wang-CSMA committed:
Command classes: Remove duplicate code used for defunct text ui

Text based UI for Aurora is now defunct and replaced by JavaFX
GUI. Therefore, defunct and unused code should be removed.

Unify handle and handleGui methods into singular handle method.

Unifying handle and handleGui improves code readability and
quality.

J-wang-CSMA committed:
Reorder all imports to fit checkstyle requirements.

J-wang-CSMA committed:
Fix generic code quality issues

Some methods and classes had missing JavaDocs comments, lines
exceeding the character limit of 120.

Identify and rectify the problems listed above.

J-wang-CSMA committed:
Remove unused methods and variables

Improves code readability and quality.

J-wang-CSMA committed:
Remove and reduce method length and magic literals

Use of magic literals can confuse others as to what
they actually represent. Long methods are hard to debug.

Replace magic literals with constants. Shorten long methods
by replacing repeated string messages with constants.

Changes will improve code quality as per the requirements
in this week's textbook content.

J-wang-CSMA committed:
Refactor long methods to make them shorter.

Long methods can be very difficult to debug and thus more
short methods that encapsulates components of the long
functions are preferred.

Compartmentalize long methods into shorter methods.

J-wang-CSMA committed:
Add Assertion statements to verify handling

It will be confusing to the user if the commands themselves were not
being executed but no feedback is given to the user on the JavaFX Gui.

Use assertions to root out potentially not executing code even with
the correct inputs.

J-wang-CSMA committed:
Merge pull request #1 from J-wang-CSMA/branch-A-CodeQuality

Improve Code Quality using Code Quality standard provided.

J-wang-CSMA committed:
Merge branch 'branch-A-Assertions'

Merging branch with main branch.

J-wang-CSMA committed:
Add DoAfter command feature and support.

Often, users might want to keep track of a task they wish to perform
either after another task or after a specific date and time. This commit
allows them to do so with the Aurora application.

Add support for DoAfter command.

Users can use the doafter command in 2 ways:
* doafter {description of task} /after {other task's tasknumber}
* doafter {description of task} /after {date and time}

J-wang-CSMA committed:
Add alert for if a Task associated with a doAfter is deleted.

If a task associated with a doAfter task is deleted, if there
is no visual feedback about it, users can get confused when they
try to find the task again.

Edit getTaskListString method to incorporate new feature.

J-wang-CSMA committed:
Add alert for if a Task associated with a doAfter is deleted

If a task associated with a doAfter task is deleted, if there
is no visual feedback about it, users can get confused when they
try to find the task again.

Edit getTaskListString method to incorporate new feature.

J-wang-CSMA committed:
Merge branch 'master' of https://github.com/J-wang-CSMA/ip

J-wang-CSMA committed:
Fix bugs related to DoAfter command.

Many bugs were present in the previous commit, such as issues with
nesting, storage and retrieval from storage.

Fix nesting by simply removing reference to task when the task is
deleted from the taskList and updating methods so they do not
recursively call downwards when they encounter a DoAfter without
an associated task.

Fix storage and retrieval by allowing for population of generated
taskList during the retrieval phase and updating of storage syntax
for DoAfters.

J-wang-CSMA committed:
Improve code quality by reducing arrowhead code and shortening methods

J-wang-CSMA committed:
Edit javadocs to fit coding standard

J-wang-CSMA committed:
Add representative image of Ui

J-wang-CSMA committed:
Update user guide for Aurora

J-wang-CSMA committed:
Edit to allow images to render in website

J-wang-CSMA committed:
Update user guide for better readability

J-wang-CSMA committed:
Edit code to make application non-resizable

J-wang-CSMA committed:
Fix naming according to Java code convention

J-wang-CSMA committed:
Resize window so Mac users can use the application without having to open a new window

J-wang-CSMA committed:
Change images to fit current ui

J-wang-CSMA committed:
Update README.md

J-wang-CSMA committed:
Remove unnecessary storage file

J-wang-CSMA committed:
Update gradle build file

J-wang-CSMA committed:
Update README.md

J-wang-CSMA committed:
Update README.md
